src/main/java/apoc/algo/algorithms/Algorithm.java

```diff
@@ -0,0 +1,205 @@
+package apoc.algo.algorithms;
+
+import org.neo4j.graphdb.ResourceIterator;
+import org.neo4j.graphdb.Result;
+import org.neo4j.kernel.internal.GraphDatabaseAPI;
+import org.neo4j.logging.Log;
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+
+public class Algorithm {
+
+    public static final int INITIAL_ARRAY_SIZE=100_000;
+    private final GraphDatabaseAPI db;
+    private final Log log;
+    private final ExecutorService pool;
+
+    public int nodeCount;
+    public int relCount;
+    public long readNodeMillis, readRelationshipMillis;
+
+    // Arrays to hold the graph.
+    // Mapping from Algo node ID to Graph nodeID
+    public int [] nodeMapping;
+
+    // Degree of each algo node.
+    public int [] sourceDegreeData;
+
+    // Starting index in relationship arrays for each algo node.
+    public int [] sourceChunkStartingIndex;
+
+    // Storing relationships
+    public int [] relationshipTarget;
+    public int [] relationshipWeight;
+
+    public Algorithm(GraphDatabaseAPI db,
+                     ExecutorService pool,
+                     Log log) {
+        this.pool = pool;
+        this.db = db;
+        this.log = log;
+    }
+
+    public boolean readNodesAndRelCypherWeighted(String relCypher, String nodeCypher) {
+        return readNodeAndRelCypherIntoArrays(relCypher, nodeCypher, true);
+    }
+
+    public boolean readNodeAndRelCypherIntoArrays(String relCypher, String nodeCypher, boolean weighted) {
+        Result nodeResult = db.execute(nodeCypher);
+
+        long before = System.currentTimeMillis();
+        ResourceIterator<Object> resultIterator = nodeResult.columnAs("id");
+        int index = 0;
+        int totalNodes = 0;
+        nodeMapping = new int[INITIAL_ARRAY_SIZE];
+        int currentSize = INITIAL_ARRAY_SIZE;
+        while(resultIterator.hasNext()) {
+            int node  = ((Long)resultIterator.next()).intValue();
```

Comment:
Actually I was trying to save on the memory by using ints for ints. For these algorithms, we were aiming at graphs with a few hundred million nodes.

If I have to get the int out of the long array, then I will have to do `intValue()`